File: 1220-count-vowels-permutation/1220-count-vowels-permutation.cpp
#define ll long long
class Solution {
public:
    const int mod = 1e9 + 7;
    int dp[20001][6];
    ll dfs(int cur, int n){
        if(n == 0)return 1;
        if(dp[n][cur] != -1)return dp[n][cur];
        if(cur == 1)return dp[n][cur] = dfs(2,n-1) % mod;
        if(cur == 2)return dp[n][cur] = (dfs(1,n-1)%mod + dfs(3,n-1)%mod)%mod;
        if(cur == 3)return dp[n][cur] = (dfs(1,n-1)%mod + dfs(2,n-1)%mod + dfs(4,n-1)%mod + dfs(5,n-1)%mod)%mod;
        if(cur == 4)return dp[n][cur] = (dfs(3,n-1)%mod + dfs(5,n-1)%mod);
        return dp[n][cur] = dfs(1,n-1)%mod; 
    }
    int countVowelPermutation(int n) {
        memset(dp,-1,sizeof(dp));
        int ans = 0;
        for(int i=1;i<=5;i++){
            ans = (ans%mod + dfs(i,n-1)%mod)%mod;
        }
        return ans;
    }
};

File: 1220-count-vowels-permutation/1220-count-vowels-permutation.py
class Solution:
    def countVowelPermutation(self, n: int) -> int:
        
        dp = [[-1]*(6) for i in range(n+1)]
        def dfs(i,c):
            if i == 0:
                return 1
            if dp[i][c] != -1:
                return dp[i][c]
            x = 0
            if c == 1:
                x = dfs(i-1,2)
            elif c == 2:
                x = dfs(i-1,3) + dfs(i-1,1)
            elif c == 3:
                x = dfs(i-1,1) + dfs(i-1,2) + dfs(i-1,4) + dfs(i-1,5)
            elif c == 4:
                x = dfs(i-1,3) + dfs(i-1,5)
            else:
                x = dfs(i-1,1)
            dp[i][c] = x
            return x
        ans = 0
        for i in range(1,6):
            ans += dfs(n-1,i)
        return ans % (10 ** 9 + 7)
        